ckanext/datastore/backend/postgres.py

```diff
@@ -1025,6 +1116,33 @@ def alter_table(context, data_dict, plugin_data):
             identifier(f['id']),
             f['type']))
 
+    # (canada fork only): foreign keys
+    # TODO: upstream contrib!!
+    foreign_keys = data_dict.get('foreign_keys')
+    if foreign_keys:
+        for f_table_name, column_name_map in foreign_keys.items():
+            constraint_name = _generate_constraint_name(f_table_name, column_name_map.keys(), column_name_map.values())
+            if constraint_name in current_foreign_constraints:
+                log.debug('Foreign key constraint {0} already exists, skipping...'.format(constraint_name))
+                continue
+            log.debug('Trying to build foreign key constraint {0} for column(s) {1}({2}) on foreign column(s) {3}({4}) ON DELETE RESTRICT ON UPDATE CASCADE'.format(
+                identifier(constraint_name),
+                identifier(data_dict['resource_id']),
+                ','.join(identifier(k) for k in column_name_map.keys()),
+                identifier(f_table_name),
+                ','.join(identifier(v) for v in column_name_map.values())
+            ))
+            #TODO: allow for deletion and update enum??
+            # might be able to do this just for upstream. for now, we can force restrict delete and cascade update.
+            # a = no action, r = restrict, c = cascade, n = set null, d = set default
+            alter_sql.append('ALTER TABLE {0} ADD CONSTRAINT {1} FOREIGN KEY ({2}) REFERENCES {3}({4})'.format(
+                identifier(data_dict['resource_id']),
+                identifier(constraint_name),
+                ','.join(identifier(k) for k in column_name_map.keys()),
+                identifier(f_table_name),
+                ','.join(identifier(v) for v in column_name_map.values())
+            ))
+
```

Comment:
where do we drop constraints that aren't passed?

Reply:
this should be done now, and we can test more heavily when doing upstream contrib.